sdks/python/apache_beam/runners/common.py

```diff
@@ -527,11 +734,11 @@ def __init__(self,
         signature.is_stateful_dofn())
     self.user_state_context = user_state_context
     self.is_splittable = signature.is_splittable_dofn()
-    self.watermark_estimator = self.signature.get_watermark_estimator()
-    self.watermark_estimator_param = (
-        self.signature.process_method.watermark_estimator_arg_name
-        if self.watermark_estimator else None)
-    self.threadsafe_restriction_tracker = None  # type: Optional[iobase.ThreadsafeRestrictionTracker]
+    self.threadsafe_restriction_tracker = None
+    self.threadsafe_watermark_estimator = None
+    # The lock which guarantee synchronization for both
+    # ThreadsafeRestrictionTracker and ThreadsafeWatermarkEstimator.
+    self._synchronized_lock = threading.Lock()
```

Comment:
IMHO making this lock external to both classes, and having some methods that do locking (exclusive of holding the lock) and others that require holding the lock, introduces complexity that is not worth the benefit (presumably making sure the watermark is never slightly stale).